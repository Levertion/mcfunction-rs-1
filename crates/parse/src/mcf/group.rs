use util::commands::Index;

#[derive(Debug, Copy, Clone, PartialEq, Eq)]
pub enum McGroupType {
    File,
    CommandNode(Index),
    Command,
    // NBT groups
    NbtCompound,
    NbtSequence,
    NbtNumber,
    NbtString,
    NbtBoolean,
    // NBT Keywords
    NbtSuffixB,
    NbtSuffixS,
    NbtSuffixL,
    NbtSuffixF,
    NbtSuffixD,
    NbtPrefixB,
    NbtPrefixI,
    NbtPrefixL,
    // NBT Compound
    NbtCompoundEntry,

    NbtPath,
    NbtPathSegment,
    NbtPathIndex,

    // Selector groups
    Selector,
    SelectorArgument,
    SelectorArgumentEntry,
    SelectorArgumentMap,
    SelectorArgumentMapEntry,
    // Selector keywords
    SelectorModP,
    SelectorModA,
    SelectorModR,
    SelectorModS,
    SelectorModE,

    // Block states
    BlockState,
    BlockStateArguments,

    ItemStack,
    ItemPredicate,

    Comment,

    Function,

    JsonObject,
    JsonObjectEntry,
    JsonList,
    JsonNull,

    Integer,
    Float,
    UnquotedString,
    ResourceLocation,
    Range,
    Uuid,
    Time,

    TimeS,
    TimeT,
    TimeD,

    Coord,
    CoordPart,

    // Keywords
    BooleanTrue,
    BooleanFalse,

    FloatSciExpUpper,
    FloatSciExpLower,

    // Special error type that signifies an error
    Error,
}
