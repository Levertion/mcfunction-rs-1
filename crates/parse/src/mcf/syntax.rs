use crate::{syntax::TokenKind, TokenSet};
use std::fmt;

#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash)]
#[repr(u8)]
pub enum McTokenKind {
    // Single Character Punctuation
    Comma = 0,
    Dot,
    Colon,
    Semicolon,
    At,
    Excl,
    Eq,
    Lt,
    Gt,
    Slash,
    Tilde,
    Caret,
    Plus,
    Dash,
    Hash,
    // Double Character Punctuation
    DotDot,
    Lte,
    Gte,
    Swap,
    AddAssign,
    SubAssign,
    MulAssign,
    DivAssign,
    ModAssign,
    // Delimiters
    LBracket,
    RBracket,
    LCurly,
    RCurly,
    // Arbitrary Length
    QuotedString, // Does not have to be terminated, and will end at the end of any line
    Word,         // String of letters or underscores
    Whitespace,   // Non linebreak whitespace
    LineBreak,    // LF, CRLF, CR
    Digits,
    // Other
    Invalid, // Used to mark a lexing error
    Eof,     // Used to mark the end of a file
}

impl TokenKind for McTokenKind {
    const WHITESPACE: Self = Self::Whitespace;
    const EOF: Self = Self::Eof;
    const LINE_BREAK: Self = Self::LineBreak;
    const WORD: Self = Self::Word;
    const DELIMITERS: TokenSet<Self> = TokenSet::empty();
}

impl From<McTokenKind> for u8 {
    fn from(it: McTokenKind) -> Self {
        it as u8
    }
}

impl fmt::Display for McTokenKind {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        use McTokenKind::*;

        write!(
            f,
            "{}",
            match self {
                Comma => ",",
                Dot => ".",
                Colon => ":",
                Semicolon => ";",
                At => "@",
                Excl => "!",
                Eq => "=",
                Lt => "<",
                Gt => ">",
                Hash => "#",
                Slash => "/",
                Tilde => "~",
                Caret => "^",
                Plus => "+",
                Dash => "-",
                DotDot => "..",
                Lte => "<=",
                Gte => ">=",
                Swap => "><",
                AddAssign => "+=",
                SubAssign => "-=",
                MulAssign => "*=",
                DivAssign => "/=",
                ModAssign => "%=",
                LBracket => "[",
                RBracket => "]",
                LCurly => "{",
                RCurly => "}",
                QuotedString => "Quoted String",
                Word => "Word",
                Whitespace => "Whitespace",
                LineBreak => "Linebreak",
                Digits => "Digits",
                Invalid => "Invalid",
                Eof => "EOF",
            }
        )
    }
}
