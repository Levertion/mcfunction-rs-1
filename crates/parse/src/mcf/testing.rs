use crate::{
    ast::{AstNode, SyntaxKind},
    mcf::{McGroupType, McTokenKind, McfLang},
    Token,
};

use std::fmt::Write;

pub fn format_astnode(node: &AstNode<McfLang>, indlevel: usize) -> String {
    let ind = "    ".repeat(indlevel);
    let mut out = String::new();
    match node.kind() {
        SyntaxKind::Token(tk) => {
            if let McTokenKind::Whitespace | McTokenKind::LineBreak = tk {
                write!(
                    out,
                    "{}Token({:?}) `{}` at {}\n",
                    ind,
                    tk,
                    node.string().escape_debug(),
                    node.span()
                )
                .unwrap();
            } else {
                write!(
                    out,
                    "{}Token({:?}) `{}` at {}\n",
                    ind,
                    tk,
                    node.string(),
                    node.span()
                )
                .unwrap();
            }
        }
        SyntaxKind::Group(gt) => {
            write!(
                out,
                "{}Group({}) at {} {{\n",
                ind,
                match gt {
                    McGroupType::CommandNode(ind) => format!("CommandNode({})", usize::from(*ind)),
                    v => format!("{:?}", v),
                },
                node.span()
            )
            .unwrap();
            for x in node.children() {
                write!(out, "{}", format_astnode(x, indlevel + 1)).unwrap();
            }
            write!(out, "{}}}\n", ind).unwrap();
        }
        SyntaxKind::Joined(gt) => {
            write!(
                out,
                "{}Joined({:?}) `{}` at {}\n",
                ind,
                gt,
                node.string(),
                node.span()
            )
            .unwrap();
            for x in node.children() {
                if let SyntaxKind::Error(err) = x.kind() {
                    write!(out, "{}- Error `{}`\n", ind, err).unwrap()
                }
            }
        }
        SyntaxKind::Error(err) => {
            write!(out, "{}Error `{}` at {}\n", ind, err, node.span()).unwrap();
        }
        SyntaxKind::Root => {
            for x in node.children() {
                write!(out, "{}", format_astnode(x, indlevel)).unwrap();
            }
        }
    };
    out
}

pub fn format_sk_list(tokens: Vec<Token<McTokenKind>>, src: &str) -> String {
    let mut out = String::new();
    for tk in tokens {
        match tk.kind() {
            McTokenKind::Whitespace | McTokenKind::LineBreak => write!(
                out,
                "{:?} `{}` at {}\n",
                tk.kind(),
                tk.string(src).escape_debug(),
                tk.span(),
            ),
            _ => write!(
                out,
                "{:?} `{}` at {}\n",
                tk.kind(),
                tk.string(src),
                tk.span(),
            ),
        }
        .unwrap();
    }
    out
}
