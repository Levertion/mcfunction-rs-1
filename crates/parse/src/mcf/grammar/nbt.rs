use super::*;
use crate::{
    mcf::{
        group::McGroupType::{self, *},
        syntax::McTokenKind::*,
        McParser,
    },
    parser::StartInfo::{self, Join, Skip},
    tokenset,
};

pub const NBT_NUMBER_SUFFIX: &[(&str, McGroupType)] = &[
    ("b", NbtSuffixB),
    ("s", NbtSuffixS),
    ("l", NbtSuffixL),
    ("f", NbtSuffixF),
    ("d", NbtSuffixD),
];

pub const NBT_SEQ_PREFIX: &[(&str, McGroupType)] =
    &[("B", NbtPrefixB), ("I", NbtPrefixI), ("L", NbtPrefixL)];

pub fn value(p: &mut McParser) {
    if p.at(QuotedString) {
        p.bump();
    } else if p.at(LCurly) {
        compound(p);
    } else if p.at(LBracket) {
        let mk = p.start(NbtSequence, Skip);

        p.bump();
        if p.nth(1) == Semicolon {
            if !p.expect_keyword(NBT_SEQ_PREFIX) {
                p.bump();
            }
            p.bump();
        }
        if !p.at(RBracket) {
            loop {
                value(p);
                if !p.eat(Comma) {
                    break;
                }
            }
        }
        p.expect(RBracket);
        p.finish(mk);
    } else if p.at(Word) && !tokenset!(p.nth(1) => ALLOWED_UQ_STRING) && p.at_keyword(BOOLEAN) {
        let mk = p.start(NbtBoolean, Join);
        p.bump();
        p.finish(mk);
    } else {
        let nmk = p.start(NbtNumber, StartInfo::None);

        if !p.try_token(float_tk, Float) {
            p.cancel(nmk);
            uq_string(p);
        } else {
            p.eat_keyword(NBT_NUMBER_SUFFIX);
            if p.at_tokens(ALLOWED_UQ_STRING) {
                p.cancel(nmk);
                uq_string(p);
            } else {
                p.finish(nmk);
            }
        }
    }
}

pub fn compound(p: &mut McParser) {
    let cpdmk = p.start(NbtCompound, Skip);

    p.bump();
    if !p.at(RCurly) {
        loop {
            let enmk = p.start(NbtCompoundEntry, Skip);
            string(p);
            p.expect(Colon);
            value(p);
            p.finish(enmk);
            if !p.eat(Comma) {
                break;
            }
        }
    }
    p.expect(RCurly);
    p.finish(cpdmk);
}

pub fn path(p: &mut McParser) {
    let mk = p.start(NbtPath, StartInfo::None);
    let mut start = true;
    loop {
        let vmk = p.start(NbtPathSegment, StartInfo::None);
        if p.at(LBracket) {
            let indmk = p.start(NbtPathIndex, StartInfo::Skip);
            p.bump();
            if !p.at(RBracket) && !p.try_token(integer_tk, Integer) {
                compound(p);
            }
            p.expect(RBracket);
            p.finish(indmk);
        } else {
            if !start {
                if !(p.eat(Dot) || p.eat(DotDot)) {
                    p.cancel(vmk);
                    break;
                }
            } else {
                start = false;
            }
            if !p.eat(Word) && !p.eat(QuotedString) {
                p.error(NbtPathSegment);
                p.finish(vmk);
                break;
            }
        }
        p.finish(vmk);
    }
    p.finish(mk);
}

#[cfg(test)]
mod tests {
    use crate::mcf::{parse, testing::format_astnode};

    use insta::assert_snapshot;

    macro_rules! nbt_test {
        ($name:ident, $e:expr) => {
            #[test]
            fn $name() {
                assert_snapshot!(parse_nbt($e));
            }
        };
    }

    macro_rules! path_test {
        ($name:ident, $e:expr) => {
            #[test]
            fn $name() {
                assert_snapshot!(parse_path($e));
            }
        };
    }

    fn parse_nbt(i: &str) -> String {
        format_astnode(&parse(i, super::value), 0)
    }

    fn parse_path(i: &str) -> String {
        format_astnode(&parse(i, super::path), 0)
    }

    nbt_test!(unsigned_int, "123");
    nbt_test!(signed_int, "-2147483648");
    nbt_test!(suffix_int, "16s");
    nbt_test!(invalid_suffix_int, "1a");

    nbt_test!(compound_empty, "{}");
    nbt_test!(compound_simple, "{foo:123}");
    nbt_test!(compound_multi, "{foo:123,bar:420}");
    nbt_test!(compound_spacing, "{\tfoo  :1564 ,  \t   bar:420   }");
    nbt_test!(compound_unclosed, "{foo:123,bar:420");
    nbt_test!(compound_nested, "{foo:{bar:{baz:\"ikr\"}}}");

    nbt_test!(list_empty, "[]");
    nbt_test!(array_empty, "[B;]");
    nbt_test!(list_simple, "['hello']");
    nbt_test!(array_simple, "[L;123456789]");
    nbt_test!(list_multi, "['hello', 123, true]");
    nbt_test!(list_nested, "[[123],[[],456,[789]]]");
    nbt_test!(list_unclosed, "[1, 2, 3, 4, 5, 6");

    nbt_test!(boolean_true, "true");
    nbt_test!(boolean_false, "false");

    nbt_test!(float_unsigned, "0.5772156649");
    nbt_test!(float_no_trail, "645423.");
    nbt_test!(float_sci, "2.2e10");
    nbt_test!(float_plus, "+2.718281828");
    nbt_test!(float_dash, "-0.61803398875");
    nbt_test!(float_no_prec, ".37412");
    nbt_test!(float_only_dot, ".");

    path_test!(empty, "");
    path_test!(single_field, "foo");
    path_test!(multi_field, "path.to.field");
    path_test!(start_index, "[0]");
    path_test!(chained_index, "[0][34][12553]");
    path_test!(mixed, "foo[0].bar.baz[1]");
    path_test!(filter, "filter[{me:123}]");
}
