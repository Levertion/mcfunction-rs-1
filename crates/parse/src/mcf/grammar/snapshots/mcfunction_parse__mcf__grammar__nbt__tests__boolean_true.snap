---
source: crates/parse/src/mcf/grammar/nbt.rs
expression: "parse_nbt(\"true\")"
---
Joined(NbtBoolean) `true` at 0:0 - 0:4

